doc/CGROUP_DELEGATION.md

```diff
@@ -0,0 +1,420 @@
+      systemd. In this case systemd will just manage the unit for you and nothing else.
+
+   b. Instead you register the service or scope through
+      `systemd-machined`. This mini-daemon is basically just a proxy for the
+      same operations as in a. The main benefit of this: this way you let the
+      system know that what you are registering is a container, and this opens
+      up certain additional integration points. For example, `journalctl -M`
+      can then be used to directly look into any container's journal logs
+      (should the container run systemd inside), or `systemctl -M` can be used
+      to directly invoke systemd operations inside the containers. Moreover
+      tools like "ps" can then show you to which container a process belongs
+      (`ps -eo pid,comm,machine`), and even gnome-system-monitor supports it.
+
+2. The *i-like-islands* option. If all you care about is your own cgroup tree,
+   and you want to have to do as little as possible with systemd and no
+   interest in integration with the rest of the system, then this is a valid
+   option. For this all you have to do is turn on `Delegate=` for your main
+   manager daemon. Then figure out the cgroup systemd placed your daemon in:
+   you can now freely create sub-cgroups beneath it. Don't forget the
+   *no-processes-in-inner-nodes* rule however: you have to move your main
+   daemon process out of that cgroup (and into a sub-cgroup) before you can
+   start further processes in any of your sub-cgroups.
+
+3. The *i-like-continents* option. In this option you'd leave your manager
+   daemon where it is, and would not turn on delegation on its unit. However,
+   as first thing you register a new scope unit with systemd, and that scope
+   unit would have `Delegate=` turned on, and then you place all your
+   containers underneath it. From systemd's PoV there'd be two units: your
+   manager service and the big scope that contains all your containers in one.
+
+BTW: if for whatever reason you say "I hate D-Bus, I'll never call any D-Bus
+API, kthxbye", then options #1 and #3 are not available, as they generally
+involve talking to systemd from your program code, via D-Bus. You still have
+option #2 in that case however, as you can simply set `Delegate=` in your
+service's unit file and you are done and have your own sub-tree. In fact, #2 is
+the one option that allows you to completely ignore systemd's existence: you
+can entirely generically follow the single rule that you just use the cgroup
+you are started in, and everything below it, whatever that might be. That said,
+maybe if you dislike D-Bus and systemd that much, the better approach might be
+to work on that, and widen your horizon a bit. You are welcome.
+
+## Controller Support
+
+systemd supports a number of controllers, but not all. Specifically:
+
+* on cgroupsv1: `cpu`, `cpuacct`, `blkio`, `memory`, `devices`, `pids`
+* on cgroupsv2: `cpu`, `io`, `memory`, `pids`
```

Comment:
So if a new cgroupsv2 controller is supported in the kernel, but not in systemd (directly), can it be made to delegate it to a scope/service with Delegate=yes? Because maybe the thing running inside that container (perhaps newer systemd) will know what to do with it? Will that situation be completely unsupported, and we'll basically have to ensure that both kernel and systemd know about a container before it can be used?

Reply:
controllers not supported by systemd can't be delegated by it, and this is unlikely to change...

The issue is that to people to delegate a controller it needs to be enabled all the way up the tree, but systemd can't do that if it doesn't know...

There has been some work in providing enough info in /sys somewhere so that userspace could delegate controllers it doesn't know at all, but systemd doesn't support that and I am not to overly keen to make that work, because we couldn't pass bitmasks around anymore, but would always have to deal with string list when talking about container masks...